### PR TITLE
resolve ambiguities between cat methods, reenable Core+Base ambiguity detection

### PR DESCRIPTION
**Pull request by Sacha0**
... specifically between `cat` as in `git cat-file` and other `cat` methods. Not certain whether and if so how testing this fix is possible: The ambiguous methods are
```julia
cat(repo::Base.LibGit2.GitRepo, spec) in Base.LibGit2 at libgit2/libgit2.jl:500
```
with
```julia
cat(catdims, xs::Union{Array{T,1} where T, Array{T,2} where T, Base.LinAlg.AbstractTriangular{T,A} where A<:(Array{T,2} where T) where T, Hermitian{T,A} where A<:(Array{T,2} where T) where T, Symmetric{T,A} where A<:(Array{T,2} where T) where T}...) in Base.SparseArrays at sparse/sparsevector.jl:969

cat(catdims, Xin::Union{Array{T,1} where T, Array{T,2} where T, Base.LinAlg.AbstractTriangular{T,A} where A<:(Array{T,2} where T) where T, Base.LinAlg.AbstractTriangular{T,A} where A<:Union{Bidiagonal, Diagonal, SparseMatrixCSC, SparseVector, SymTridiagonal, Tridiagonal} where T, Bidiagonal, Diagonal, Hermitian{T,A} where A<:(Array{T,2} where T) where T, Hermitian{T,A} where A<:Union{Bidiagonal, Diagonal, SparseMatrixCSC, SparseVector, SymTridiagonal, Tridiagonal} where T, SparseMatrixCSC, SparseVector, SymTridiagonal, Symmetric{T,A} where A<:(Array{T,2} where T) where T, Symmetric{T,A} where A<:Union{Bidiagonal, Diagonal, SparseMatrixCSC, SparseVector, SymTridiagonal, Tridiagonal} where T, Tridiagonal}...) in Base.SparseArrays at sparse/sparsevector.jl:939

cat(catdims, A::AbstractArray{T,N} where N...) where T in Base at abstractarray.jl:1267

cat(catdims, xs::Union{Array{T,1}, Array{T,2}, Base.LinAlg.AbstractTriangular{T,A} where A<:(Array{T,2} where T), Hermitian{T,A} where A<:(Array{T,2} where T), Symmetric{T,A} where A<:(Array{T,2} where T)}...) where T in Base.SparseArrays at sparse/sparsevector.jl:976
```
with intersections
```julia
Tuple{Base.#cat,Base.LibGit2.GitRepo,Union{Array{T,1} where T, Array{T,2} where T, Base.LinAlg.AbstractTriangular{T,A} where A<:(Array{T,2} where T) where T, Hermitian{T,A} where A<:(Array{T,2} where T) where T, Symmetric{T,A} where A<:(Array{T,2} where T) where T}}

Tuple{Base.#cat,Base.LibGit2.GitRepo,Union{Array{T,1} where T, Array{T,2} where T, Base.LinAlg.AbstractTriangular{T,A} where A<:(Array{T,2} where T) where T, Base.LinAlg.AbstractTriangular{T,A} where A<:Union{Bidiagonal, Diagonal, SparseMatrixCSC, SparseVector, SymTridiagonal, Tridiagonal} where T, Bidiagonal, Diagonal, Hermitian{T,A} where A<:(Array{T,2} where T) where T, Hermitian{T,A} where A<:Union{Bidiagonal, Diagonal, SparseMatrixCSC, SparseVector, SymTridiagonal, Tridiagonal} where T, SparseMatrixCSC, SparseVector, SymTridiagonal, Symmetric{T,A} where A<:(Array{T,2} where T) where T, Symmetric{T,A} where A<:Union{Bidiagonal, Diagonal, SparseMatrixCSC, SparseVector, SymTridiagonal, Tridiagonal} where T, Tridiagonal}}

Tuple{Base.#cat,Base.LibGit2.GitRepo,AbstractArray{T,N} where N} where T

Tuple{Base.#cat,Base.LibGit2.GitRepo,Union{Array{T,1}, Array{T,2}, Base.LinAlg.AbstractTriangular{T,A} where A<:(Array{T,2} where T), Hermitian{T,A} where A<:(Array{T,2} where T), Symmetric{T,A} where A<:(Array{T,2} where T)}} where T
```
Calling `cat` with appropriate arguments to hit the ambiguity yields, without this fix, a `MethodError` for the ambiguity and, with the fix, also a `MethodError` (due to absence of a method for the necessarily(?) nonsensical argument combination). Thoughts? Thanks!

### CODE REVIEW

**Comment by tkelman** [comment]
I think this one's subject to future deprecation, but PkgDev may have been using it which might need cleaning up

**Comment by Sacha0** [comment]
> I think this one's subject to future deprecation, but PkgDev may have been using it which might need cleaning up

Not certain I follow you. Is the implication that merging without a test is fine, deprecation anticipated? Thanks!

**Comment by Sacha0** [comment]
This pull request now fixes all remaining ambiguities in `Core` and `Base`, so I added a commit re-enabling the ambiguity detection test. Best!

**Comment by Sacha0** [comment]
(Insofar as I can tell, none of this CI round's failures are related to the content of this PR?)

**Comment by timholy** [comment]
Looks like you were hoping for review from @kshyatt? If she can't do it, and you're confident the failures are unrelated, I'd say go ahead and merge.

**Comment by tkelman** [comment]
I think the failures are related. In https://travis-ci.org/JuliaLang/julia/jobs/210305433 (backed up to https://gist.github.com/6bc06858a64b0b88b866c4ba9a739caa) and https://ci.appveyor.com/project/JuliaLang/julia/build/1.0.12094/job/koxad7y72yrkyyg1, the ambiguity test never finished and CI timed out waiting for it.

File: test/ambiguous.jl [hunk 1]
```diff
@@ -135,7 +135,7 @@ ambs = detect_ambiguities(Ambig5)
 # Test that Core and Base are free of ambiguities
 # TODO jb/subtype: we now detect a lot more
 @test !isempty(detect_ambiguities(Core, Base; imported=true, ambiguous_bottom=true))
```
Comment (tkelman) [inline test/ambiguous.jl]: should maybe have a comment here that this being non-empty isn't actually required, but is likely to be the case

**Comment by Sacha0** [comment]
Apologies for being off-grid the last few days, and thanks for rebasing / rerunning CI! Seems all went well this round? Best!

**Comment by StefanKarpinski** [comment]
Yes, this time it seems all good, so I think we can squash and merge this.

**Comment by Sacha0** [comment]
Thanks all!

**Comment by tkelman** [comment]
We may need to revert this. Froze on the centos 7 buildbot, started running the ambiguity test but didn't finish - https://build.julialang.org/builders/build_centos7.1-x64/builds/51/steps/shell_2/logs/stdio

**Comment by tkelman** [comment]
The Furlong type from TestHelpers also has ambiguities, which only cause a failure when dimensionful, ambiguous, and some other test that triggers the ambiguity run on the same worker. `make testall1` gives the full list.

**Comment by timholy** [comment]
Run it in a fresh worker? (One that hasn't run anything else.)

**Comment by tkelman** [comment]
That doesn't really fix `male testall1` unless it always runs first. it might not help with the freezing on the buildbots earlier, but could be tried.

**Comment by timholy** [comment]
We used to issue text warnings (which had a lot of problems) while julia was being built, and this provided very quick feedback about problematic changes. Running it first is the closest analog; in a multiprocess test run, perhaps it should run even before the workers are started.

**Comment by StefanKarpinski** [comment]
Running it before any tests seems sensible.

**Pull request merged**